### PR TITLE
Added support for mobile_installation.log files

### LINKED ISSUE #4048: Parser for iOS Mobile Installation Logs
Create a parser to parser iOS mobile installation logs. They are found at

```
/private/var/installd/Library/Logs/MobileInstalation/.log
```

### PR DESCRIPTION
**Pull request by rick-slin**
## One line description of pull request
Adding support to parse mobile installation logs for iOS.

## Description:

**Related issue (if applicable):** fixes #4048

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

**Comment by joachimmetz** [comment]
@rick-slin thx for the proposed changes, the CI tests are currently failing because the changes do not match the style guide https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html

**Comment by joachimmetz** [comment]
assigning @jleaniz for a first review

File: plaso/parsers/__init__.py [hunk 1]
```diff
@@ -35,6 +35,7 @@
 from plaso.parsers import mac_wifi
 from plaso.parsers import mactime
 from plaso.parsers import mcafeeav
+from plaso.parsers import mobile_installation
```
Comment (jleaniz) [inline plaso/parsers/__init__.py:38]: I would suggest to use a more specific module name here, such as ios_mobile_installation_logs or ios_mobile_installation

File: plaso/parsers/mobile_installation.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+"""Parser for iOS mobile installation logs files obtained from iOS device.
+Based on the research from Alex Brignoni."""
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:3]: Could you please add a URL referencing the research? I found a GitHub repository but maybe you are referncing something else.

File: plaso/parsers/mobile_installation.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class MobileInstallationEventData(events.EventData):
+  """Mobile Installation log event data
+  Attributes:
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:19]: Please add a description for all the attributes and sort the attributes alphabetically as per the style guide:
https://github.com/log2timeline/l2tdocs/blob/main/process/Style-guide.md

File: plaso/parsers/mobile_installation.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    number (str):
+    severity (str):
+    id (str):
+    originator (str): process that created the entry
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:23]: I suggest changing originator == originating_process for more clarity.

File: plaso/parsers/mobile_installation.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    self.severity = None
+    self.id = None
+    self.originator = None
+    self.body = None
+
+
+class MobileInstallationParser(text_parser.PyparsingMultiLineTextParser):
+  """Parser for iOS Mobile Installation logs."""
+
+  NAME = 'mobile_installation.log'
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:42]: mobile_installation.log ==> ios_mobile_installation.log

File: plaso/parsers/mobile_installation.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            'Nov': 11,
+            'Dec': 12}
+
+  ONE_OR_TWO_DIGITS = text_parser.PyparsingConstants.ONE_OR_TWO_DIGITS
+  FOUR_DIGITS = text_parser.PyparsingConstants.FOUR_DIGITS
+  THREE_LETTERS = text_parser.PyparsingConstants.THREE_LETTERS
+  TIME_ELEMENTS = text_parser.PyparsingConstants.TIME_ELEMENTS
+
+  _TIMESTAMP = (THREE_LETTERS.suppress() + THREE_LETTERS('month') +
+                ONE_OR_TWO_DIGITS('day') + TIME_ELEMENTS + FOUR_DIGITS('year'))
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:64]: style nit: use four spaces for hanging indents throughout (https://github.com/log2timeline/l2tdocs/blob/main/process/Style-guide.md#indentation)

See: https://github.com/log2timeline/plaso/blob/main/plaso/parsers/aws_elb_access.py#L129 as an example of this.

File: plaso/parsers/mobile_installation.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+  _ELEMENT_NUMBER = pyparsing.Suppress('[') + \
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:66]: style nit: no need for the extra \ character here and in the other pyparsing expressions below.

File: plaso/parsers/mobile_installation.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class MobileInstallationEventData(events.EventData):
+  """Mobile Installation log event data
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:18]: style nit: add new line here.

**Review by jleaniz — CHANGES_REQUESTED** [review]
@rick-slin Thanks for the new parser. I left some comments from my initial review, mostly style related. Please reach out here or on the Open Source DFIR Slack plaso channel if you have any questions.

File: plaso/parsers/mobile_installation.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  Attributes:
+    number (str):
+    severity (str):
+    id (str):
```
Comment (jleaniz) [inline plaso/parsers/mobile_installation.py:22]: id == identifier

File: plaso/parsers/mobile_installation.py (hunk 2 above)
Comment (rick-slin) [reply plaso/parsers/mobile_installation.py:3]: Got mixed up in my references.

File: data/formatters/ios.yaml [hunk 10]
```diff
@@ -78,3 +78,11 @@ message:
 short_message:
 - 'Name: {name}'
 - 'Message: {text}'
+---
+type: 'conditional'
+data_type: 'mobile_installation:line'
+message:
+- 'Originator: {originator}'
```
Comment (jleaniz) [inline data/formatters/ios.yaml:85]: This should be updated to reflect the new attribute name and data_type name

**Comment by jleaniz** [comment]
@rick-slin Thanks for making the changes. Since these are also IOS related logs that you are parsing, please prefix the data type name with ios: as you did with the sysdiagnose parser, just so that it is clearer what platform it is for.

File: data/formatters/ios.yaml [hunk 11]
```diff
@@ -80,6 +80,14 @@ short_message:
 - 'Message: {text}'
 ---
 type: 'conditional'
+data_type: 'ios:mobile_installation:line'
+message:
+- 'Originating call: {originating_call}'
+- 'Body: {body}'
+short_message:
+  - 'Body: {body}'
```
Comment (joachimmetz) [inline data/formatters/ios.yaml:88]: consistency nit: remove indentation

File: plaso/parsers/__init__.py [hunk 12]
```diff
@@ -36,6 +36,7 @@
 from plaso.parsers import mac_wifi
 from plaso.parsers import mactime
 from plaso.parsers import mcafeeav
+from plaso.parsers import ios_mobile_installation_logs
```
Comment (joachimmetz) [inline plaso/parsers/__init__.py:39]: style nit: make sure these are in alphabetical ordering

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
+"""Parser for iOS mobile installation logs files obtained from iOS device.
+Based on the research by Alex Brignoni.
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:3]: Please don't put links of blog posts in source, these go out of date very quickly. Instead please create a forensicwiki article.

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class MobileInstallationEventData(events.EventData):
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:19]: Please prefix with IOS

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  """Mobile Installation log event data
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:20]: "Mobile Installation log event data" => "iOS mobile installation log event data."

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+  Attributes:
+    process_identifier (str): process_identifier
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:23]: consistency nit: sort these alphabetically

Comment (joachimmetz) [reply plaso/parsers/ios_mobile_installation_logs.py:23]: add a dot at the end of each of them.

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    originating_call (str): call that created the entry
+    body (str): body of the event line
+  """
+
+  DATA_TYPE = 'ios:mobile_installation:line'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(MobileInstallationEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.process_identifier = None
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:34]: consistency nit: sort these alphabetically

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+  _TIMESTAMP = (THREE_LETTERS.suppress() + THREE_LETTERS('month') +
+      ONE_OR_TWO_DIGITS('day') + TIME_ELEMENTS + FOUR_DIGITS('year'))
+
+  _ELEMENT_NUMBER = (pyparsing.Suppress('[') +
+      pyparsing.Word(pyparsing.nums)('process_identifier') +
+      pyparsing.Suppress(']'))
+
+  _ELEMENT_SEVERITY = (pyparsing.Suppress('<') +
+      pyparsing.Word(pyparsing.alphanums)('severity') + pyparsing.Suppress('>'))
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:73]: consistency/readability nit: please don't use shorthand.

File: plaso/parsers/ios_mobile_installation_logs.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+        and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+
```
Comment (joachimmetz) [inline plaso/parsers/ios_mobile_installation_logs.py:107]: remove empty line

**Comment by joachimmetz** [comment]
@rick-slin could you make the requested changes please, most are nits for code consistency.

**Review by jleaniz — APPROVED** [review]

**Pull request merged**